modules/tinymce/src/core/test/ts/browser/content/EditorContentEventsTest.ts

```diff
@@ -123,4 +123,28 @@ describe('browser.tinymce.core.content.EditorContentEventsTest', () => {
     assertEvents([]);
     TinyAssertions.assertContent(editor, '<p>Selection content</p>');
   });
+
+  const data = { hello: 'world', test: 132 };
+  Arr.each([ 'BeforeSetContent', 'SetContent' ], (action) => {
+    it(`TINY-9143: Can pass custom data object to "${action}" event`, () => {
+      const editor = hook.editor();
+      editor.setContent('<p>initial</p>');
+      editor.once(action, (e) => {
+        assert.equal(data.hello, e.hello);
+        assert.equal(data.test, e.test);
+      });
+      editor.setContent('<p>new</p>', data);
```

Comment:
This will pass even if the event never fires since the assert is inside the event handler. It would be more safe something like this:

```suggestion
      const editor = hook.editor();
      const lastEventState = Singleton.value<SetContentEvent>();
      editor.setContent('<p>initial</p>');
      editor.once(action, (e) => {
        lastEventState.set(e);
      });
      editor.setContent('<p>new</p>', data);
      const lastEvent = lastEvent.get().getOrDie('Should be set');
      assert.equal(lastEvent.hello, data.hello);
      assert.equal(lastEvent.test, data.test);
```

Or maybe use a let and null not since it's just test code we could be a bit more relaxed. I also flipped the assert.equal since I think it's (actual, expected) in chai.

Reply:
`Singleton` solution is clear and readable, thanks a lot! Updated [here](https://github.com/tinymce/tinymce/pull/8215/commits/f1d9ca09b1fea6a3c05996853cedcce6b7bd489b)